crypto/dh/dh_lib.c

```diff
@@ -246,14 +246,21 @@ void DH_get0_pqg(const DH *dh, BIGNUM **p, BIGNUM **q, BIGNUM **g)
 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 {
     /* q is optional */
-    if (p == NULL || g == NULL)
+    if ((p == NULL && dh->p == NULL)
+        || (g == NULL && dh->g == NULL))
```

Comment:
From the last comments on openssl-dev, I got the impression that it might be good if those set0 functions would be seen as initializers, which means it would be an error to try to initialize fields that were already initialized.  That's why I proposed the expression I proposed.  Your suggest, @mattcaswell, doesn't quite cover that aspect.